### PR TITLE
feat: new home page, blog engine, blog layouts and routing engine

### PR DESCRIPTION
**Pull request by ovflowd**
## Description

This PR introduces several changes on the `next-data` pieces of the Node.js Website, including:

- Refactor on Promise-chains of `next-data` Providers
- Refactor of Blog Data Generators, Providers and Fetch methods
- Introduced new Pagination System for the Blog
  - The old Pagination System (years) became dedicated categories, so that they're still accessible on the web
- Refactored how we allow dynamic URLs to be generated (DYNAMIC_ROUTES)
- Cleanup on Blog-related code
- Cleanup on constants (including Sentry constants, which are now separated to a dedicated file)
- Introduced the Base Blog Category Layout
- Fixed a few components (such as Preview, and Blog Cards)
  - Preview component content now resizes by the usage of Container Queries
- Introduced the Blog Post Layout and Content Layout
- Made adjustments on some Components and styles for some Layouts
- Redesigned the Home Page based on feedback from the TSC

### Screenshots

Screenshot 2024-01-09 at 18 02 14

Screenshot 2024-01-09 at 18 02 08

Screenshot 2024-01-09 at 18 00 43

### Branch Preview

https://nodejs-org-git-feat-new-blog-pagination-system-openjs.vercel.app

## Validation

Blog Pages should work as expected, Pagination should work as expected, with the caveats that pages are now different routes.

### CODE REVIEW

File: components/MDX/CodeTabs/index.tsx
Comment (mikeesto) [inline components/MDX/CodeTabs/index.tsx:1]: Small lint error on line 37

`error TS2322: Type '{ children: Element[]; linkUrl?: string | undefined; linkText?: string | undefined; tabs: { key: string; label: string; }[]; defaultValue: string; }' is not assignable to type 'IntrinsicAttributes & Pick<TabsProps & { tabs: Tab[]; addons?: ReactNode; } & { children?: ReactNode; }, "tabs"> & { ...; } & { ...; }'.`

**Comment by ovflowd** [comment]
cc @nodejs/nodejs-website I believe this is ready for review, but it needs some intensive review, including stylistic reviews (not just how it looks, but also if the tags (tailwind) are properly used, properly indented and if we can remove tag duplication)

As usual any maintainer is free and welcome to make changes. Thanks!

File: i18n/locales/en.json
Comment (richardlau) [inline i18n/locales/en.json]: npm should not be capitalized (it's [not an acronym](https://docs.npmjs.com/policies/logos-and-usage#general-rules)).

File: pages/en/new-design/index.mdx
Comment (richardlau) [inline pages/en/new-design/index.mdx]:
```suggestion
          <small>Node.js {release.version} with long-term support and npm {release.npm}</small>
```
LTS stands for "long-term support".

npm should not be capitalized (it's [not an acronym](https://docs.npmjs.com/policies/logos-and-usage#general-rules)).

File: pages/en/new-design/index.mdx
Comment (richardlau) [inline pages/en/new-design/index.mdx]: npm should not be capitalized (it's [not an acronym](https://docs.npmjs.com/policies/logos-and-usage#general-rules)).

Comment (ovflowd) [reply pages/en/new-design/index.mdx]: Oh

File: components/Common/CodeTabs/index.module.css
Comment (AugustinMauroy) [inline components/Common/CodeTabs/index.module.css]:
```suggestion
        @apply size-4
```

File: pages/en/new-design/index.mdx
Comment (AugustinMauroy) [inline pages/en/new-design/index.mdx:19]: It's a cool change that's been made. But I don't think it's a good idea to have the same two buttons. What Hayden has done in the design is more intelligent, putting two buttons with different styles creates an asymmetry that highlights one. I'm not an expert in design, I'm just learning. I think asking Hayden's opinion wouldn't be a bad idea.

Comment (ovflowd) [reply pages/en/new-design/index.mdx:19]: Right, and I agree with you. These are experimental changes, because after all who approves these designs and content changes is the TSC, not Hayden.

So as much as I value his opinions and vision here, in the end what the TSC decides is what should be done.

There are still ongoing discussions regarding if we should have two buttons or not and there is no clear consensus yet.

**Comment by ovflowd** [comment]
@nodejs/nodejs-website I believe I've finished iterating from feedback from the @nodejs/distros 

We can finally review this Pull Request.

File: pages/en/new-design/index.mdx
Comment (AugustinMauroy) [inline pages/en/new-design/index.mdx]:
```suggestion
    const json = await res.json();
```
Line it's to long it's create a layout shift

https://github.com/nodejs/nodejs.org/assets/97875033/b0a72ca9-215e-439e-a932-ddcb3e2172c4

File: styles/new/effects.css
Comment (AugustinMauroy) [inline styles/new/effects.css]:
```suggestion
    size-full
```

File: pages/en/new-design/index.mdx
Comment (ovflowd) [reply pages/en/new-design/index.mdx]: It's not a layout shift. It's just that different screen sizes will have different way of rendering this.

Comment (ovflowd) [reply pages/en/new-design/index.mdx]: So if your screen is bigger or smaller you will have "line breaks"

Comment (AugustinMauroy) [reply pages/en/new-design/index.mdx]: Wouldn't it be better to give it a precise height?

Comment (ovflowd) [reply pages/en/new-design/index.mdx]: We can't do that.

File: pages/en/new-design/index.mdx
Comment (bmuenzenmeyer) [inline pages/en/new-design/index.mdx:90]: this doesn't seem related to the rest of the code and I'd suggest we delete - unless I am missing something about the intent of this tab or the internals of `fetch`
```suggestion
```

File: i18n/locales/en.json
Comment (bmuenzenmeyer) [inline i18n/locales/en.json:122]: `common.pagination` covered these same keys on lines 136 thru 142

File: package.json
Comment (bmuenzenmeyer) [inline package.json:130]: surprised to see this removed and not causing linting issues - i didn't think it was resolved yet 🤷 

https://github.com/simonsmith/stylelint-selector-bem-pattern/pull/68

File: pages/en/new-design/index.mdx
Comment (bmuenzenmeyer) [inline pages/en/new-design/index.mdx:1]: this file needs internationalization, right?

**Review by bmuenzenmeyer — COMMENTED** [review]
looking good so far - only had time for a quick pass

File: pages/en/new-design/index.mdx
Comment (ovflowd) [reply pages/en/new-design/index.mdx:1]: Yes, but its a markdown file. People will translate it directly by making ie: pages/fr/index.mdx

Comment (bmuenzenmeyer) [reply pages/en/new-design/index.mdx:1]: duh, silly me

File: package.json
Comment (ovflowd) [reply package.json:130]: IDK either 🤷

File: i18n/locales/en.json
Comment (ovflowd) [reply i18n/locales/en.json:122]: Right but this one is the legacy Next/Previous which is going to be removed once we remove the old website codebase.

File: pages/en/new-design/index.mdx
Comment (ovflowd) [reply pages/en/new-design/index.mdx:90]: No no, these are two different examples of stream. One using fetch and one using fs

File: i18n/locales/en.json
Comment (bmuenzenmeyer) [reply i18n/locales/en.json:122]: Then why change it now? 😁

Comment (ovflowd) [reply i18n/locales/en.json:122]: Please look carefully at the changes. These strings are used on the legacy blog pagination system.

And as you can see this PR overhauled the pagination system, meaning on the legacy blog the next and previous got switched positions, hence I had to remove the "|" character that was on the en.json file (tbh these sort of separators shouldnt be in a lang file) otherwise it would look like "| Next >< Previous" instead of "< Next Previous >" hmm just please compare main and the branch and youl see the difference xD

**Comment by canerakdas** [comment]
According to the Lighthouse report, there were two minor problems with the blog detail page in mobile resolution, I updated them;

- Links must be distinguishable without relying on color [MDX Anchors]
  - https://dequeuniversity.com/rules/axe/4.8/link-in-text-block
- Tap targets are not sized appropriately [Metabar Anchors]
  - https://developer.chrome.com/docs/lighthouse/seo/tap-targets/

Apart from these two situations, the only thing that caught my attention during my checks is that the blog post layout is not centered. We can center this layout, but the background gradient is cut when it exceeds the maximum content width. So I'm not sure about this solution 🤔 

image

File: pages/en/new-design/index.mdx
Comment (bmuenzenmeyer) [reply pages/en/new-design/index.mdx:90]: haha i find that confusing but that's just me perhaps

**Review by bmuenzenmeyer — APPROVED** [review]
Let's keep this moving - momentum is building!

**Comment by bmuenzenmeyer** [comment]
A tiny design nit is that the `Copy to Clipboard` button is very prominent in the smaller coding snippets, which I find distracting from the content. This should be a de-emphasized control in my opinion.

image


Compare to the homepage which uses a de-emphasized version

image

File: components/Common/ProgressionSidebar/index.tsx
Comment (shanpriyan) [inline components/Common/ProgressionSidebar/index.tsx:32]:
```suggestion
          key={groupName}
```

File: components/Containers/Sidebar/index.tsx
Comment (shanpriyan) [inline components/Containers/Sidebar/index.tsx:32]:
```suggestion
          key={groupName}
```

**Review by shanpriyan — APPROVED** [review]
Just few nits. LGTM 🚀

**Comment by shanpriyan** [comment]
> A tiny design nit is that the `Copy to Clipboard` button is very prominent in the smaller coding snippets, which I find distracting from the content. This should be a de-emphasized control in my opinion.
> 
> image
> 
> Compare to the homepage which uses a de-emphasized version
> 
> image

Can we can just use the copy icon and remove the `Copy to clipboard` text ? 🤔

**Review by AugustinMauroy — APPROVED** [review]
LGTM !

File: components/Common/ProgressionSidebar/index.tsx
Comment (ovflowd) [reply components/Common/ProgressionSidebar/index.tsx:32]: .toString() needs to be called, as the type of `groupName` is `FormattedMessage` which isn't necessarily a string.

File: components/Containers/Sidebar/index.tsx
Comment (ovflowd) [reply components/Containers/Sidebar/index.tsx:32]: .toString() needs to be called, as the type of `groupName` is `FormattedMessage` which isn't necessarily a string.

**Comment by ovflowd** [comment]
> A tiny design nit is that the `Copy to Clipboard` button is very prominent in the smaller coding snippets, which I find distracting from the content. This should be a de-emphasized control in my opinion.
> 
> image
> 
> Compare to the homepage which uses a de-emphasized version
> 
> image

Is the ask here to have it using the Home Page colors?

**Comment by ovflowd** [comment]
Anyhow, @bmuenzenmeyer feel free to open an issue or PR regarding the Copy to Clipboard button as it is unrelated to this PR. It already became convoluted enough 😅

File: pages/en/new-design/index.mdx
Comment (ovflowd) [reply pages/en/new-design/index.mdx:90]: It might be actually. We can revisit this in a follow-up PR!

**Pull request merged**
